diff --git a/.gitignore b/.gitignore
index d1c6979e76..24402e70d7 100644
--- a/.gitignore
+++ b/.gitignore
@@ -25,3 +25,5 @@ nfapi_nr_interface_scf
 *.log
 *.out
 CMakeUserPresets.json
+common/utils/T/tracer/t_tracer_app_gnb
+common/utils/T/tracer/t_tracer_app_ue
diff --git a/openair2/COMMON/f1ap_messages_types.h b/openair2/COMMON/f1ap_messages_types.h
index e44e1dccb4..d2ef5d3762 100644
--- a/openair2/COMMON/f1ap_messages_types.h
+++ b/openair2/COMMON/f1ap_messages_types.h
@@ -86,6 +86,9 @@
 /* 9.3.1.42 of 3GPP TS 38.473 - gNB-CU System Information */
 #define F1AP_MAX_NO_SIB_TYPES 32
 
+/* 9.3.1.86 of 3GPP TS 38.473 - Additional SIB Message List */
+#define PWS_MAX_NO_Additional_SIBs 63
+
 typedef struct f1ap_net_config_t {
   char *CU_f1_ip_address;
   char *DU_f1c_ip_address;
@@ -213,6 +216,31 @@ typedef struct f1ap_setup_resp_s {
 
 } f1ap_setup_resp_t;
 
+typedef struct sib_msg_s {
+  /// RRC container with system information owned by gNB-CU
+  uint8_t *SI_container;
+  int SI_container_length;
+} sib_msg_t;
+
+typedef struct pws_system_information_s{
+  /// SIB type (6, 7, 8)
+  int sib_type;
+  /// First SIB message segment 
+  sib_msg_t sib_message;
+  /// Additional SIB message list (for segmentation) 
+  int num_additional_sibs;
+  sib_msg_t additional_sibs[PWS_MAX_NO_Additional_SIBs];
+
+} pws_system_information_t;
+
+typedef struct write_replace_warning_req_s{
+  ///PWS System information
+  pws_system_information_t pws_system_information;
+  // NR Global Cell Id
+  uint64_t nr_cellid;
+  
+} write_replace_warning_req_t;
+
 typedef struct f1ap_gnb_cu_configuration_update_s {
   /// Transaction ID
   uint64_t transaction_id;
diff --git a/openair2/LAYER2/NR_MAC_gNB/config.c b/openair2/LAYER2/NR_MAC_gNB/config.c
index ed4d902345..9454e7f254 100644
--- a/openair2/LAYER2/NR_MAC_gNB/config.c
+++ b/openair2/LAYER2/NR_MAC_gNB/config.c
@@ -960,12 +960,88 @@ bool nr_mac_configure_other_sib(gNB_MAC_INST *nrmac, int num_cu_sib, const f1ap_
   // generate otherSIB payloads
   cc->other_sib_bcch_length[0] = encode_sysinfo_ie(sysInfo, cc->other_sib_bcch_pdu[0], sizeof(cc->other_sib_bcch_pdu[0]));
   cc->other_sib_bcch_length[1] = encode_sysinfo_ie(sysInfov17, cc->other_sib_bcch_pdu[1], sizeof(cc->other_sib_bcch_pdu[1]));
+  cc->other_si_used += 2;
 
   ASN_STRUCT_FREE(asn_DEF_NR_SystemInformation_IEs, sysInfo);
   ASN_STRUCT_FREE(asn_DEF_NR_SystemInformation_IEs, sysInfov17);
   return true;
 }
 
+void nr_mac_configure_pws_si(gNB_MAC_INST *nrmac, const pws_system_information_t *pws_system_information)
+{
+  
+  NR_COMMON_channels_t *cc = &nrmac->common_channels[0];
+
+
+  int num_segments = 1 + pws_system_information->num_additional_sibs;
+
+  NR_SystemInformation_IEs_t **sysInfo_list = calloc(num_segments, sizeof(*sysInfo_list));
+  for (int i = 0; i < num_segments; i++)
+    sysInfo_list[i] = calloc(1, sizeof(*sysInfo_list[i]));
+
+
+  uint8_t *SI_container;
+  int SI_container_length;
+  for (int i = 0; i < num_segments; i++) {
+    
+    if(i == 0) {
+      SI_container = pws_system_information->sib_message.SI_container;
+      SI_container_length = pws_system_information->sib_message.SI_container_length;
+    }
+    else {
+      SI_container = pws_system_information->additional_sibs[i-1].SI_container;
+      SI_container_length = pws_system_information->additional_sibs[i-1].SI_container_length;
+    }
+    
+    struct NR_SystemInformation_IEs__sib_TypeAndInfo__Member *type = calloc(1, sizeof(*type));
+    type->present = NR_SystemInformation_IEs__sib_TypeAndInfo__Member_PR_sib8;
+    // SIB8 coming from CU need to decode it
+    NR_SIB8_t *sib8 = NULL;
+    asn_dec_rval_t dec_rval = uper_decode(NULL,
+                                          &asn_DEF_NR_SIB8,
+                                          (void **)&sib8,
+                                          SI_container,
+                                          SI_container_length,
+                                          0,
+                                          0);
+    if (dec_rval.code != RC_OK) {
+      LOG_E(NR_MAC, "cannot decode SIB8 from CU\n");
+      free(SI_container);
+    }
+    type->choice.sib8 = sib8;
+    add_sib_to_systeminformation(sysInfo_list[i], type);      
+  }
+
+  const int start = cc->other_si_used;
+  const int old_count = cc->pws_si_count;
+
+
+  update_SIB1_PWS_SI(cc->sib1, num_segments, pws_system_information->sib_type, old_count);
+  cc->sib1_bcch_length = encode_SIB_NR(cc->sib1, cc->sib1_bcch_pdu, sizeof(cc->sib1_bcch_pdu));
+  AssertFatal(cc->sib1_bcch_length > 0, "could not encode SIB1\n");
+
+
+  // generate SI messages PDU payloads
+  for (int i = 0; i < num_segments; i++) {
+    const int si_idx = start + i;
+    AssertFatal(si_idx < MAX_SI_MESSAGE, "PWS SI index %d out of range\n", si_idx);
+
+    cc->other_sib_bcch_length[si_idx] = encode_sysinfo_ie(sysInfo_list[i], 
+                                                       cc->other_sib_bcch_pdu[si_idx], 
+                                                       sizeof(cc->other_sib_bcch_pdu[si_idx]));
+
+    ASN_STRUCT_FREE(asn_DEF_NR_SystemInformation_IEs, sysInfo_list[i]);
+  }
+  free(sysInfo_list);
+  for (int i = num_segments; i < old_count; i++) {
+    const int si_idx = start + i;
+    cc->other_sib_bcch_length[si_idx] = 0;
+  }
+
+  cc->pws_si_count = num_segments;
+  
+}
+
 bool nr_update_sib19(const gnb_sat_position_update_t *sat_position)
 {
   gNB_MAC_INST *nrmac = RC.nrmac[0];
diff --git a/openair2/LAYER2/NR_MAC_gNB/gNB_scheduler_bch.c b/openair2/LAYER2/NR_MAC_gNB/gNB_scheduler_bch.c
index 6dd4886911..d71ba64099 100644
--- a/openair2/LAYER2/NR_MAC_gNB/gNB_scheduler_bch.c
+++ b/openair2/LAYER2/NR_MAC_gNB/gNB_scheduler_bch.c
@@ -855,7 +855,7 @@ void schedule_nr_other_sib(module_id_t module_idP,
                                         rel_slot[ssb]))
         continue;
 
-      other_sib_sched_control(module_idP, frame, slot, ssb, ss, DL_req, TX_req, 0);
+      other_sib_sched_control(module_idP, frame, slot, ssb, ss, DL_req, TX_req, i);
     }
     if (!schedInfo17)
       continue;
diff --git a/openair2/LAYER2/NR_MAC_gNB/mac_proto.h b/openair2/LAYER2/NR_MAC_gNB/mac_proto.h
index a8d4c76b31..a717873eb0 100644
--- a/openair2/LAYER2/NR_MAC_gNB/mac_proto.h
+++ b/openair2/LAYER2/NR_MAC_gNB/mac_proto.h
@@ -60,6 +60,7 @@ void nr_mac_send_f1_setup_req(void);
 void nr_mac_config_scc(gNB_MAC_INST *nrmac, NR_ServingCellConfigCommon_t *scc, const nr_mac_config_t *mac_config);
 void nr_mac_configure_sib1(gNB_MAC_INST *nrmac, const plmn_id_t *plmn, uint64_t cellID, int tac);
 bool nr_mac_configure_other_sib(gNB_MAC_INST *nrmac, int num_cu_sib, const f1ap_sib_msg_t cu_sib[num_cu_sib]);
+void nr_mac_configure_pws_si(gNB_MAC_INST *nrmac, const pws_system_information_t *pws_system_information);
 bool nr_mac_add_test_ue(gNB_MAC_INST *nrmac, uint32_t rnti, NR_CellGroupConfig_t *CellGroup);
 void nr_mac_prepare_ra_ue(gNB_MAC_INST *nrmac, NR_UE_info_t *UE);
 bool add_new_UE_RA(gNB_MAC_INST *nr_mac, NR_UE_info_t *UE);
diff --git a/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.c b/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.c
index c271824e75..1d216e7094 100644
--- a/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.c
+++ b/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.c
@@ -217,6 +217,21 @@ void f1_setup_failure(const f1ap_setup_failure_t *failure)
   exit(1);
 }
 
+void write_replace_warning_req(const write_replace_warning_req_t *wrw)
+{ 
+  LOG_I(MAC, "received Write Replace Warning Request from CU\n");
+
+  const pws_system_information_t *pws_system_information = &wrw->pws_system_information;
+
+  gNB_MAC_INST *mac = RC.nrmac[0];
+  NR_SCHED_LOCK(&mac->sched_lock);
+
+  nr_mac_configure_pws_si(mac, pws_system_information);
+
+  NR_SCHED_UNLOCK(&mac->sched_lock);
+  
+}
+
 void gnb_du_configuration_update_acknowledge(const f1ap_gnb_du_configuration_update_acknowledge_t *ack)
 {
   (void)ack;
diff --git a/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.h b/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.h
index 1ba31f8e80..951f66748d 100644
--- a/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.h
+++ b/openair2/LAYER2/NR_MAC_gNB/mac_rrc_dl_handler.h
@@ -30,6 +30,7 @@ void f1_reset_cu_initiated(const f1ap_reset_t *reset);
 void f1_reset_acknowledge_du_initiated(const f1ap_reset_ack_t *ack);
 void f1_setup_response(const f1ap_setup_resp_t *resp);
 void f1_setup_failure(const f1ap_setup_failure_t *failure);
+void write_replace_warning_req(const write_replace_warning_req_t *req);
 void gnb_du_configuration_update_acknowledge(const f1ap_gnb_du_configuration_update_acknowledge_t *ack);
 NR_CellGroupConfig_t *clone_CellGroupConfig(const NR_CellGroupConfig_t *orig);
 void ue_context_setup_request(const f1ap_ue_context_setup_req_t *req);
diff --git a/openair2/LAYER2/NR_MAC_gNB/nr_mac_gNB.h b/openair2/LAYER2/NR_MAC_gNB/nr_mac_gNB.h
index 5914c74a45..814ce3e866 100644
--- a/openair2/LAYER2/NR_MAC_gNB/nr_mac_gNB.h
+++ b/openair2/LAYER2/NR_MAC_gNB/nr_mac_gNB.h
@@ -304,8 +304,10 @@ typedef struct {
   uint8_t sib1_bcch_pdu[NR_MAX_SIB_LENGTH / 8];
   int sib1_bcch_length;
   /// used for otherSIB data
-  uint8_t other_sib_bcch_pdu[2][NR_MAX_SIB_LENGTH / 8];
-  int other_sib_bcch_length[2];
+  int other_si_used;
+  int pws_si_count;
+  uint8_t other_sib_bcch_pdu[MAX_SI_MESSAGE][NR_MAX_SIB_LENGTH / 8];
+  int other_sib_bcch_length[MAX_SI_MESSAGE];
   /// VRB map for common channels
   uint16_t vrb_map[MAX_NUM_BEAM_PERIODS][275];
   /// VRB map for common channels and PUSCH, dynamically allocated because
diff --git a/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.c b/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.c
index 50f5fd684c..0e03d8a719 100644
--- a/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.c
+++ b/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.c
@@ -2675,6 +2675,43 @@ void update_SIB1_NR_SI(NR_BCCH_DL_SCH_Message_t *sib1_bcch, int num_sibs, int si
   }
 }
 
+void update_SIB1_PWS_SI(NR_BCCH_DL_SCH_Message_t *sib1_bcch, int num_segments, int type, int old_num_segments)
+{
+  NR_SIB1_t *sib1 = sib1_bcch->message.choice.c1->choice.systemInformationBlockType1;
+
+  NR_SI_SchedulingInfo_t *info = sib1->si_SchedulingInfo;
+  if (info == NULL) {
+    LOG_I(MAC, "No other SI has been used\n");
+    info = calloc(1, sizeof(*info));
+    AssertFatal(info != NULL, "calloc failed\n");
+    info->si_WindowLength = NR_SI_SchedulingInfo__si_WindowLength_s40;
+    sib1->si_SchedulingInfo = info;
+  }
+
+  while (old_num_segments-- > 0 && info->schedulingInfoList.list.count > 0) {
+    const int idx = info->schedulingInfoList.list.count - 1;
+    NR_SchedulingInfo_t *e = info->schedulingInfoList.list.array[idx];
+
+    ASN_STRUCT_FREE(asn_DEF_NR_SchedulingInfo, e);
+
+    info->schedulingInfoList.list.array[idx] = NULL;
+    info->schedulingInfoList.list.count--;
+  }
+
+  for (int i = 0; i < num_segments; i++) {
+      NR_SchedulingInfo_t *schedulingInfo_sib8 = calloc(1, sizeof(*schedulingInfo_sib8));
+      schedulingInfo_sib8->si_BroadcastStatus = NR_SchedulingInfo__si_BroadcastStatus_broadcasting;
+      schedulingInfo_sib8->si_Periodicity = NR_SchedulingInfo__si_Periodicity_rf64;
+      
+      NR_SIB_TypeInfo_t *mapping = calloc(1, sizeof(*mapping));
+      mapping->type = type - 2; 
+      
+      asn1cSeqAdd(&schedulingInfo_sib8->sib_MappingInfo.list, mapping);
+      asn1cSeqAdd(&info->schedulingInfoList.list, schedulingInfo_sib8);
+  }
+
+}
+
 int encode_sysinfo_ie(NR_SystemInformation_IEs_t *sysInfo, uint8_t *buf, int len)
 {
   if (sysInfo->sib_TypeAndInfo.list.count == 0)
diff --git a/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.h b/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.h
index 6752db8bc8..ea6c3be641 100644
--- a/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.h
+++ b/openair2/LAYER2/NR_MAC_gNB/nr_radio_config.h
@@ -75,12 +75,16 @@ int encode_MeasurementTimingConfiguration(const struct NR_MeasurementTimingConfi
 void free_MeasurementTimingConfiguration(struct NR_MeasurementTimingConfiguration *mtc);
 
 #define NR_MAX_SIB_LENGTH 2976 // 3GPP TS 38.331 section 5.2.1
+#define MAX_SI_MESSAGE 32 // 3GPP TS 38.331 section 6.4
+
 NR_BCCH_DL_SCH_Message_t *get_SIB1_NR(const NR_ServingCellConfigCommon_t *scc,
                                       const plmn_id_t *plmn,
                                       uint64_t cellID,
                                       int tac,
                                       const nr_mac_config_t *mac_config);
 void update_SIB1_NR_SI(NR_BCCH_DL_SCH_Message_t *sib1, int num_sibs, int sibs[num_sibs]);
+void update_SIB1_PWS_SI(NR_BCCH_DL_SCH_Message_t *sib1_bcch, int num_segments, int type, int old_num_segments);
+
 int encode_sysinfo_ie(NR_SystemInformation_IEs_t *sysInfo, uint8_t *buf, int len);
 void free_SIB1_NR(NR_BCCH_DL_SCH_Message_t *sib1);
 int encode_SIB_NR(NR_BCCH_DL_SCH_Message_t *sib, uint8_t *buffer, int max_buffer_size);
diff --git a/openair2/RRC/NR/MESSAGES/asn1_msg.c b/openair2/RRC/NR/MESSAGES/asn1_msg.c
index 585b1e48e2..85d1bc0ac9 100644
--- a/openair2/RRC/NR/MESSAGES/asn1_msg.c
+++ b/openair2/RRC/NR/MESSAGES/asn1_msg.c
@@ -279,6 +279,448 @@ int do_SIB2_NR(uint8_t **msg_SIB2, NR_SSB_MTC_t *ssbmtc)
   ASN_STRUCT_FREE(asn_DEF_NR_SIB2, sib2);
   return size;
 }
+//------------------------------------------------------------------------------
+int get_config_value(const char *filepath, const char *key, char *out) {
+    FILE *file = fopen(filepath, "r");
+    if (!file)
+        return 0;
+    
+    size_t line_buf_size = (strcmp(key, "text") == 0) ? 4000 : 30;
+
+    char *line = malloc(line_buf_size);
+
+    if (!line) {
+        fclose(file);
+        return 0;
+    }
+
+    size_t keyLen = strlen(key);
+
+    while (fgets(line, (int)line_buf_size, file)) {
+          if (strncmp(line, key, keyLen) == 0){            
+            char *value = line + keyLen + 1;
+
+            char *newline = strchr(value, ';');
+            if (newline) *newline = '\0';
+
+            strcpy(out, value);
+            for (char *p = out; *p; p++) {
+              if (*p == '|') *p = '\n';
+            }
+            fclose(file);
+            free(line);
+            return 1; 
+          }
+
+    }
+    fclose(file);
+    free(line);
+    return 0;
+}
+
+void hex2bytes(const char *hex, uint8_t *out, int len) {
+  for (int i = 0; i < len; ++i) {
+    sscanf(hex + 2*i, "%2hhx", &out[i]);
+  }
+}
+
+static unsigned char gsm_7bit_alphabet[] = {
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0a, 0xff, 0xff, 0x0d, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x20, 0x21, 0x22, 0x23, 0x02, 0x25, 0x26, 0x27,
+    0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x3b,
+    0x3c, 0x3d, 0x3e, 0x3f, 0x00, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f,
+    0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x3c, 0x2f, 0x3e, 0x14, 0x11, 0xff, 0x61, 0x62, 0x63,
+    0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b, 0x6c, 0x6d, 0x6e, 0x6f, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
+    0x78, 0x79, 0x7a, 0x28, 0x40, 0x29, 0x3d, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0c, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x5e, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0x40, 0xff, 0x01, 0xff, 0x03, 0xff, 0x7b, 0x7d, 0xff, 0xff, 0xff, 0xff, 0xff, 0x5c, 0xff, 0xff, 0xff, 0xff, 0xff,
+    0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x5b, 0x7e, 0x5d, 0xff, 0x7c, 0xff, 0xff, 0xff, 0xff, 0x5b, 0x0e, 0x1c, 0x09,
+    0xff, 0x1f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x5d, 0xff, 0xff, 0xff, 0xff, 0x5c, 0xff, 0x0b, 0xff, 0xff, 0xff,
+    0x5e, 0xff, 0xff, 0x1e, 0x7f, 0xff, 0xff, 0xff, 0x7b, 0x0f, 0x1d, 0xff, 0x04, 0x05, 0xff, 0xff, 0x07, 0xff, 0xff, 0xff,
+    0xff, 0x7d, 0x08, 0xff, 0xff, 0xff, 0x7c, 0xff, 0x0c, 0x06, 0xff, 0xff, 0x7e, 0xff, 0xff};
+
+int gsm_septet_encode(uint8_t *result, const char *data)
+{
+  int i, y = 0;
+  uint8_t ch;
+  for (i = 0; i < strlen(data); i++) {
+    ch = data[i];
+    switch (ch) {
+      /* fall-through for extension characters */
+      case 0x0c:
+      case 0x5e:
+      case 0x7b:
+      case 0x7d:
+      case 0x5c:
+      case 0x5b:
+      case 0x7e:
+      case 0x5d:
+      case 0x7c:
+        result[y++] = 0x1b;
+      /* fall-through */
+      default:
+        result[y] = gsm_7bit_alphabet[ch];
+        break;
+    }
+    y++;
+  }
+
+  return y;
+}
+
+size_t gsm_septet_pack(uint8_t *result, const uint8_t *rdata, size_t septet_len, uint8_t padding)
+{
+  int i = 0;
+  size_t z = 0;
+  uint8_t cb, nb;
+  int shift = 0;
+  uint8_t *data = malloc(septet_len + 1);
+
+  if (padding) {
+    shift = 7 - padding;
+    /* the first zero is needed for padding */
+    memcpy(data + 1, rdata, septet_len);
+    data[0] = 0x00;
+    septet_len++;
+  } else
+    memcpy(data, rdata, septet_len);
+
+  for (i = 0; i < septet_len; i++) {
+    if (shift == 7) {
+      /*
+       * special end case with the. This is necessary if the
+       * last septet fits into the previous octet. E.g. 48
+       * non-extension characters:
+       *   ....ag ( a = 1100001, g = 1100111)
+       * result[40] = 100001 XX, result[41] = 1100111 1 */
+      if (i + 1 < septet_len) {
+        shift = 0;
+        continue;
+      } else if (i + 1 == septet_len)
+        break;
+    }
+
+    cb = (data[i] & 0x7f) >> shift;
+    if (i + 1 < septet_len) {
+      nb = (data[i + 1] & 0x7f) << (7 - shift);
+      cb = cb | nb;
+    }
+
+    result[z++] = cb;
+    shift++;
+  }
+
+  free(data);
+
+  return z;
+}
+
+int gsm_7bit_encode_n(uint8_t *result, size_t n, const char *data, uint8_t *octets)
+{
+  int y = 0;
+  size_t o;
+  size_t max_septets = n * 8 / 7;
+
+  /* prepare for the worst case, every character expanding to two bytes */
+  uint8_t *rdata = malloc(strlen(data) * 2);
+  y = gsm_septet_encode(rdata, data);
+
+  if (y > max_septets) {
+    /*
+     * Limit the number of septets to avoid the generation
+     * of more than n octets.
+     */
+    y = max_septets;
+  }
+
+  o = gsm_septet_pack(result, rdata, y, 0);
+
+  if (octets)
+    *octets = o;
+
+  free(rdata);
+
+  /*
+   * We don't care about the number of octets, because they are not
+   * unique. E.g.:
+   *  1.) 46 non-extension characters + 1 extension character
+   *         => (46 * 7 bit + (1 * (2 * 7 bit))) / 8 bit =  42 octets
+   *  2.) 47 non-extension characters
+   *         => (47 * 7 bit) / 8 bit = 41,125 = 42 octets
+   *  3.) 48 non-extension characters
+   *         => (48 * 7 bit) / 8 bit = 42 octects
+   */
+  return y;
+}
+
+size_t utf8_to_ucs2(const char *utf8, uint8_t *ucs2, size_t max_bytes)
+{
+  if (!ucs2)
+    return 0;
+  const unsigned char *p = (const unsigned char *)utf8;
+  size_t i = 0;
+
+  while (*p) {
+    uint32_t codepoint = 0;
+
+    if (*p < 0x80) {
+      codepoint = *p++;
+    } else if ((*p & 0xE0) == 0xC0) {
+      codepoint = (*p++ & 0x1F) << 6;
+      codepoint |= (*p++ & 0x3F);
+    } else if ((*p & 0xF0) == 0xE0) {
+      codepoint = (*p++ & 0x0F) << 12;
+      codepoint |= (*p++ & 0x3F) << 6;
+      codepoint |= (*p++ & 0x3F);
+    } else {
+      // Skip 4-byte (U+10000+), not valid in UCS-2
+      p += 4;
+      continue;
+    }
+
+    if (codepoint > 0xFFFF)
+      continue; // Out of UCS-2 range
+
+    ucs2[i++] = (uint8_t)(codepoint >> 8);
+    ucs2[i++] = (uint8_t)codepoint;
+    if(i >= max_bytes - 1) {
+      // Prevent overflow, UCS-2 buffer should not exceed max bytes
+      return i;
+    }
+  }
+
+  return i;
+}
+
+
+
+int build_sib8_segments(nr_SIBs_t **out_segments,
+                        int *out_num_segments)
+{
+  *out_segments    = NULL;
+  *out_num_segments = 0;
+
+  char hex_str[5]; 
+  const char *config_file = "../../../sib8.conf";
+  
+  char alert_buffer[4000] = {0};
+  get_config_value(config_file, "text", alert_buffer);
+
+  uint8_t *dcs = CALLOC(1, 1);
+  get_config_value(config_file, "dataCodingScheme", hex_str);
+  hex2bytes(hex_str, dcs, 1);
+
+  char mode[5];
+  get_config_value(config_file, "mode", mode);
+ 
+
+  const int MAX_SEGMENTS = 32;
+  const int BYTES_PER_PAGE = 82;
+  const int CHARS_PER_PAGE_GSM7 = 93;
+
+  uint8_t *encoded = NULL;
+  int text_len = 0;
+  size_t total_octets = 0;
+  size_t total_pages = 0;
+
+  if (*dcs == 0x01) {
+
+    text_len = strlen(alert_buffer);
+    total_pages = (text_len + CHARS_PER_PAGE_GSM7 - 1) / CHARS_PER_PAGE_GSM7;
+
+  } else if (*dcs == 0x48) {
+
+    size_t max_bytes = 3 * BYTES_PER_PAGE * MAX_SEGMENTS;
+    encoded = malloc(max_bytes);
+
+    total_octets = utf8_to_ucs2(alert_buffer, encoded, max_bytes);
+    total_pages = (total_octets + BYTES_PER_PAGE -1) / BYTES_PER_PAGE;
+    
+  } 
+       
+  size_t num_segments = 0;
+
+
+  int pages_per_segment = 1;
+  if (mode[0] == '0')
+    pages_per_segment = 3;
+
+  num_segments = (total_pages + pages_per_segment - 1) / pages_per_segment;
+  
+  if (num_segments > MAX_SEGMENTS)
+    num_segments = MAX_SEGMENTS;
+
+  nr_SIBs_t *segments = calloc(num_segments, sizeof(*segments));
+
+  for (int seg_no = 0; seg_no < num_segments; seg_no++) {
+
+    int page_base = seg_no * pages_per_segment;
+    int remaining_pages = total_pages - page_base;
+    int pages_in_this_segment = remaining_pages > pages_per_segment ? pages_per_segment : remaining_pages;
+
+    NR_SIB8_t *sib8 = calloc(1, sizeof(*sib8));
+    AssertFatal(sib8, "SIB8: out of memory\n");
+    uint8_t* segment;
+    int i = 0;
+
+    if(mode[0] == '1'){
+      
+      sib8->dataCodingScheme = CALLOC(1, sizeof(OCTET_STRING_t));
+      sib8->dataCodingScheme->size = 1;
+      sib8->dataCodingScheme->buf = CALLOC(1, 1);
+      get_config_value(config_file, "dataCodingScheme", hex_str);
+      hex2bytes(hex_str, sib8->dataCodingScheme->buf, 1);
+      
+      // First byte is for the number of segments, BYTES_PER_PAGE for the page content, last byte for page meaningful length (because a page MUST contain 82 bytes padded with zero)
+      sib8->warningMessageSegment.size = 1 + BYTES_PER_PAGE + 1;
+      sib8->warningMessageSegment.buf  = CALLOC(sib8->warningMessageSegment.size, sizeof(uint8_t));
+      
+      segment =  sib8->warningMessageSegment.buf;
+      segment[i++] = 1;
+
+      sib8->warningMessageSegmentNumber = 0;
+      sib8->warningMessageSegmentType = NR_SIB8__warningMessageSegmentType_lastSegment;
+
+      sib8->serialNumber.size = 2;
+      sib8->serialNumber.buf  = CALLOC(2, sizeof(uint8_t));
+      sib8->serialNumber.buf[0] = 0x3F;
+      sib8->serialNumber.buf[1] = seg_no;
+      sib8->serialNumber.bits_unused = 0;
+
+    } else if(mode[0] == '2'){
+      
+      if (seg_no < num_segments/2){
+
+        sib8->warningMessageSegmentType = NR_SIB8__warningMessageSegmentType_notLastSegment;
+
+        sib8->dataCodingScheme = CALLOC(1, sizeof(OCTET_STRING_t));
+        sib8->dataCodingScheme->size = 1;
+        sib8->dataCodingScheme->buf = CALLOC(1, 1);
+        get_config_value(config_file, "dataCodingScheme", hex_str);
+        hex2bytes(hex_str, sib8->dataCodingScheme->buf, 1);
+
+        sib8->warningMessageSegment.size = 1 + BYTES_PER_PAGE + 1;
+        sib8->warningMessageSegment.buf  = CALLOC(sib8->warningMessageSegment.size, sizeof(uint8_t));
+
+        segment =  sib8->warningMessageSegment.buf;
+        segment[i++] = 2;
+
+        sib8->serialNumber.size = 2;
+        sib8->serialNumber.buf  = CALLOC(2, sizeof(uint8_t));
+        sib8->serialNumber.buf[0] = 0x3F;
+        sib8->serialNumber.buf[1] = seg_no;
+        sib8->serialNumber.bits_unused = 0;
+
+        sib8->warningMessageSegmentNumber = 0;
+
+      } else {
+        sib8->warningMessageSegmentType = NR_SIB8__warningMessageSegmentType_lastSegment;
+        sib8->dataCodingScheme = NULL;
+
+        sib8->warningMessageSegment.size = BYTES_PER_PAGE + 1;
+        sib8->warningMessageSegment.buf  = CALLOC(sib8->warningMessageSegment.size, sizeof(uint8_t));
+
+        segment =  sib8->warningMessageSegment.buf;
+        sib8->serialNumber.size = 2;
+        sib8->serialNumber.buf  = CALLOC(2, sizeof(uint8_t));
+        sib8->serialNumber.buf[0] = 0x3F;
+        sib8->serialNumber.buf[1] = seg_no - num_segments/2;
+        sib8->serialNumber.bits_unused = 0;
+
+        sib8->warningMessageSegmentNumber = 1;
+      } 
+    } else {
+      if(seg_no == 0){
+          sib8->dataCodingScheme = CALLOC(1, sizeof(OCTET_STRING_t));
+          sib8->dataCodingScheme->size = 1;
+          sib8->dataCodingScheme->buf = CALLOC(1, 1);
+          get_config_value(config_file, "dataCodingScheme", hex_str);
+          hex2bytes(hex_str, sib8->dataCodingScheme->buf, 1);
+          
+          sib8->warningMessageSegment.size = 1 + pages_in_this_segment * (BYTES_PER_PAGE + 1);
+          sib8->warningMessageSegment.buf  = CALLOC(sib8->warningMessageSegment.size, sizeof(uint8_t));
+        
+          segment =  sib8->warningMessageSegment.buf;
+          segment[i++] = total_pages;
+      } else{
+          sib8->dataCodingScheme = NULL;
+        
+          sib8->warningMessageSegment.size = pages_in_this_segment * (BYTES_PER_PAGE + 1);
+          sib8->warningMessageSegment.buf  = CALLOC(sib8->warningMessageSegment.size, sizeof(uint8_t));
+        
+          segment =  sib8->warningMessageSegment.buf;
+      }
+  
+      sib8->warningMessageSegmentNumber = seg_no;
+
+      if (seg_no == num_segments - 1) {
+        sib8->warningMessageSegmentType = NR_SIB8__warningMessageSegmentType_lastSegment;
+      } else {
+        sib8->warningMessageSegmentType = NR_SIB8__warningMessageSegmentType_notLastSegment;
+      }
+    
+      sib8->serialNumber.size = 2;
+      sib8->serialNumber.buf  = CALLOC(2, sizeof(uint8_t));
+      get_config_value(config_file, "serialNumber", hex_str);
+      hex2bytes(hex_str, sib8->serialNumber.buf, 2);
+      sib8->serialNumber.bits_unused = 0;
+
+    }
+
+
+    for (int p = 0; p < pages_in_this_segment; p++) {
+      int page_idx = page_base + p;
+
+      uint8_t *dst_page = segment + i + p * (BYTES_PER_PAGE + 1);
+      uint8_t *dst_len  = dst_page + BYTES_PER_PAGE;
+
+      if (*dcs == 0x48) {
+        int start = page_idx * BYTES_PER_PAGE;
+        int remain = (int)total_octets - start;
+        int this_len = remain > BYTES_PER_PAGE ? BYTES_PER_PAGE : remain;
+        if (this_len > 0) memcpy(dst_page, encoded + start, this_len);
+        *dst_len = (uint8_t)this_len;
+      } else {
+        int start = page_idx * CHARS_PER_PAGE_GSM7;
+        gsm_7bit_encode_n(dst_page, BYTES_PER_PAGE, alert_buffer + start, dst_len);
+      }
+    }
+        
+    sib8->messageIdentifier.size = 2;
+    sib8->messageIdentifier.buf  = CALLOC(2, sizeof(uint8_t));
+    get_config_value(config_file, "messageIdentifier", hex_str);
+    hex2bytes(hex_str, sib8->messageIdentifier.buf, 2);
+    sib8->messageIdentifier.bits_unused = 0;
+   
+    LOG_I(NR_RRC, "[SIB8] segments numer:%d and number of segments:%ld total_pages: %ld\n",seg_no,num_segments, total_pages);
+
+    printf("[SIB8][HEX] ");
+    for (size_t i = 0; i < sib8->warningMessageSegment.size; i++)
+        printf("%02X ", segment[i]);
+    printf("\n");
+
+
+    // ASN.1 encode this SIB8
+    uint8_t *encoded_sib8 = NULL;
+    ssize_t enc_size = uper_encode_to_new_buffer(&asn_DEF_NR_SIB8,
+                                                 NULL,
+                                                 (void *)sib8,
+                                                 (void **)&encoded_sib8);
+    AssertFatal(enc_size > 0, "SIB8: ASN1 encoding failed (seg %d)\n", seg_no);
+
+    
+    segments[seg_no].SIB_buffer  = encoded_sib8;
+    segments[seg_no].SIB_size = enc_size;
+
+    ASN_STRUCT_FREE(asn_DEF_NR_SIB8, sib8);
+  }
+
+  free(encoded);
+
+  *out_segments    = segments;
+  *out_num_segments = num_segments;
+  return 0;
+}
 
 int do_RRCReject(uint8_t *const buffer)
 {
diff --git a/openair2/RRC/NR/MESSAGES/asn1_msg.h b/openair2/RRC/NR/MESSAGES/asn1_msg.h
index eb8da91c2a..f531ef6aec 100644
--- a/openair2/RRC/NR/MESSAGES/asn1_msg.h
+++ b/openair2/RRC/NR/MESSAGES/asn1_msg.h
@@ -90,6 +90,8 @@ int do_SIB2_NR(uint8_t **msg_SIB2, NR_SSB_MTC_t *ssbmtc);
 
 int do_RRCReject(uint8_t *const buffer);
 
+int build_sib8_segments(nr_SIBs_t **out_segments, int *out_num_segments);
+
 int do_RRCSetup(uint8_t *const buffer,
                 size_t buffer_size,
                 const uint8_t transaction_id,
diff --git a/openair2/RRC/NR/mac_rrc_dl.h b/openair2/RRC/NR/mac_rrc_dl.h
index 4bd106d679..55c4283d08 100644
--- a/openair2/RRC/NR/mac_rrc_dl.h
+++ b/openair2/RRC/NR/mac_rrc_dl.h
@@ -30,6 +30,7 @@ typedef void (*f1_reset_acknowledge_du_initiated_func_t)(sctp_assoc_t assoc_id,
 
 typedef void (*f1_setup_response_func_t)(sctp_assoc_t assoc_id, const f1ap_setup_resp_t *resp);
 typedef void (*f1_setup_failure_func_t)(sctp_assoc_t assoc_id, const f1ap_setup_failure_t *fail);
+typedef void (*write_replace_warning_req_func_t)(sctp_assoc_t assoc_id, const write_replace_warning_req_t *req);
 typedef void (*gnb_du_configuration_update_ack_func_t)(sctp_assoc_t assoc_id,
                                                        const f1ap_gnb_du_configuration_update_acknowledge_t *ack);
 
diff --git a/openair2/RRC/NR/mac_rrc_dl_direct.c b/openair2/RRC/NR/mac_rrc_dl_direct.c
index 83c1130fcb..8ca1936b12 100644
--- a/openair2/RRC/NR/mac_rrc_dl_direct.c
+++ b/openair2/RRC/NR/mac_rrc_dl_direct.c
@@ -52,6 +52,12 @@ static void f1_setup_failure_direct(sctp_assoc_t assoc_id, const f1ap_setup_fail
   f1_setup_failure(fail);
 }
 
+static void write_replace_warning_req_direct(sctp_assoc_t assoc_id, const write_replace_warning_req_t *wrw)
+{
+  AssertFatal(assoc_id == -1, "illegal assoc_id %d\n", assoc_id);
+  write_replace_warning_req(wrw);
+}
+
 static void gnb_du_configuration_update_ack_direct(sctp_assoc_t assoc_id, const f1ap_gnb_du_configuration_update_acknowledge_t *ack)
 {
   AssertFatal(assoc_id == -1, "illegal assoc_id %d\n", assoc_id);
@@ -99,6 +105,7 @@ void mac_rrc_dl_direct_init(nr_mac_rrc_dl_if_t *mac_rrc)
   mac_rrc->f1_reset = f1_reset_cu_initiated_direct;
   mac_rrc->f1_reset_acknowledge = f1_reset_acknowledge_du_initiated_direct;
   mac_rrc->f1_setup_response = f1_setup_response_direct;
+  mac_rrc->write_replace_warning_req = write_replace_warning_req_direct;
   mac_rrc->f1_setup_failure = f1_setup_failure_direct;
   mac_rrc->gnb_du_configuration_update_acknowledge = gnb_du_configuration_update_ack_direct;
   mac_rrc->ue_context_setup_request = ue_context_setup_request_direct;
diff --git a/openair2/RRC/NR/nr_rrc_defs.h b/openair2/RRC/NR/nr_rrc_defs.h
index 4d24a771c0..addbb38c91 100644
--- a/openair2/RRC/NR/nr_rrc_defs.h
+++ b/openair2/RRC/NR/nr_rrc_defs.h
@@ -319,6 +319,7 @@ typedef struct nr_mac_rrc_dl_if_s {
   f1_reset_acknowledge_du_initiated_func_t f1_reset_acknowledge;
   f1_setup_response_func_t f1_setup_response;
   f1_setup_failure_func_t f1_setup_failure;
+  write_replace_warning_req_func_t write_replace_warning_req;
   gnb_du_configuration_update_ack_func_t gnb_du_configuration_update_acknowledge;
   ue_context_setup_request_func_t ue_context_setup_request;
   ue_context_modification_request_func_t ue_context_modification_request;
diff --git a/openair2/RRC/NR/rrc_gNB.c b/openair2/RRC/NR/rrc_gNB.c
index d2f08fd4b4..f6ea8fae8a 100644
--- a/openair2/RRC/NR/rrc_gNB.c
+++ b/openair2/RRC/NR/rrc_gNB.c
@@ -3023,12 +3023,34 @@ static bool write_rrc_stats(const gNB_RRC_INST *rrc)
   return true;
 }
 
+void handle_pws_timer_expiry(){
+  const char *cfg_path = "../../../sib8.conf";
+
+  static time_t last_mtime = 0;
+  static off_t  last_size  = -1;
+
+  struct stat st;
+  if (stat(cfg_path, &st) != 0) {
+    return;
+  }
+
+  if (st.st_mtime == last_mtime && st.st_size == last_size)
+    return;
+
+  write_replace_warning_req_trigger(-1);
+  
+  last_mtime = st.st_mtime;
+  last_size  = st.st_size;
+}
+
 void *rrc_gnb_task(void *args_p) {
   MessageDef *msg_p;
   instance_t                         instance;
   int                                result;
 
   long stats_timer_id = 1;
+  long pws_timer_id = 2;
+  timer_setup(5, 0, TASK_RRC_GNB, 0, TIMER_PERIODIC, NULL, &pws_timer_id);
   if (!IS_SOFTMODEM_NOSTATS) {
     /* timer to write stats to file */
     timer_setup(1, 0, TASK_RRC_GNB, 0, TIMER_PERIODIC, NULL, &stats_timer_id);
@@ -3051,6 +3073,7 @@ void *rrc_gnb_task(void *args_p) {
       case TERMINATE_MESSAGE:
         LOG_W(NR_RRC, " *** Exiting NR_RRC thread\n");
         timer_remove(stats_timer_id);
+        timer_remove(pws_timer_id);
         itti_exit_task();
         break;
 
@@ -3062,6 +3085,8 @@ void *rrc_gnb_task(void *args_p) {
         if (TIMER_HAS_EXPIRED(msg_p).timer_id == stats_timer_id) {
           if (!write_rrc_stats(RC.nrrrc[0]))
             timer_remove(stats_timer_id);
+        } else if (TIMER_HAS_EXPIRED(msg_p).timer_id == pws_timer_id) {
+          handle_pws_timer_expiry();
         } else {
           itti_send_msg_to_task(TASK_RRC_GNB, 0, TIMER_HAS_EXPIRED(msg_p).arg); /* see rrc_gNB_process_NGAP_PDUSESSION_SETUP_REQ() */
         }
diff --git a/openair2/RRC/NR/rrc_gNB_du.c b/openair2/RRC/NR/rrc_gNB_du.c
index 246df60626..3d2bae6e9f 100644
--- a/openair2/RRC/NR/rrc_gNB_du.c
+++ b/openair2/RRC/NR/rrc_gNB_du.c
@@ -289,6 +289,55 @@ static bool valid_du_in_neighbour_configs(const seq_arr_t *neighbour_cell_config
   return true;
 }
 
+void write_replace_warning_req_trigger(sctp_assoc_t assoc_id)
+{
+  gNB_RRC_INST *rrc = RC.nrrrc[0];
+  DevAssert(rrc);
+  if (!rrc || !rrc->mac_rrc.write_replace_warning_req) {
+    LOG_E(NR_RRC, "PWS: mac_rrc.write_replace_warning_req not initialized\n");
+    return;
+  }
+  
+  write_replace_warning_req_t wrw = {
+      .pws_system_information = {
+        .sib_type = 8,
+        .num_additional_sibs = 0,
+      },
+  };
+
+
+  pws_system_information_t *pws = &wrw.pws_system_information;
+
+  nr_SIBs_t *seg_array = NULL;
+  int nseg = 0;
+  if (build_sib8_segments(&seg_array, &nseg) < 0 || nseg <= 0) {
+    AssertFatal(false, "SIB8: cannot build segments\n");
+  }
+
+  // First segment
+  pws->sib_message.SI_container        = seg_array[0].SIB_buffer;
+  pws->sib_message.SI_container_length = seg_array[0].SIB_size;
+
+  // Remaining segments (if any)
+  int add = nseg - 1;
+  if (add > PWS_MAX_NO_Additional_SIBs)
+    add = PWS_MAX_NO_Additional_SIBs;
+
+  pws->num_additional_sibs = add;
+
+  for (int i = 0; i < add; i++) {
+    pws->additional_sibs[i].SI_container        = seg_array[i + 1].SIB_buffer;
+    pws->additional_sibs[i].SI_container_length = seg_array[i + 1].SIB_size;
+  }
+    
+  rrc->mac_rrc.write_replace_warning_req(assoc_id, &wrw);
+
+  for (int i = 0; i < nseg; i++) {
+    free(seg_array[i].SIB_buffer);
+  }
+  free(seg_array);
+}
+
 void rrc_gNB_process_f1_setup_req(f1ap_setup_req_t *req, sctp_assoc_t assoc_id)
 {
   AssertFatal(assoc_id != 0, "illegal assoc_id == 0: should be -1 (monolithic) or >0 (split)\n");
@@ -442,6 +491,10 @@ void rrc_gNB_process_f1_setup_req(f1ap_setup_req_t *req, sctp_assoc_t assoc_id)
   if (rrc->node_name != NULL)
     resp.gNB_CU_name = strdup(rrc->node_name);
   rrc->mac_rrc.f1_setup_response(assoc_id, &resp);
+
+
+  write_replace_warning_req_trigger(assoc_id);
+
   free_f1ap_setup_response(&resp);
 
   /* we need to setup one default UE for phy-test and do-ra modes in the MAC */
diff --git a/openair2/RRC/NR/rrc_gNB_du.h b/openair2/RRC/NR/rrc_gNB_du.h
index 514416211f..8e05e0dc65 100644
--- a/openair2/RRC/NR/rrc_gNB_du.h
+++ b/openair2/RRC/NR/rrc_gNB_du.h
@@ -33,6 +33,7 @@ struct nr_rrc_du_container_t;
 struct f1ap_gnb_du_configuration_update_s;
 struct f1ap_served_cell_info_t;
 
+void write_replace_warning_req_trigger(sctp_assoc_t assoc_id);
 void rrc_gNB_process_f1_setup_req(struct f1ap_setup_req_s *req, sctp_assoc_t assoc_id);
 void rrc_CU_process_f1_lost_connection(struct gNB_RRC_INST_s *rrc, struct f1ap_lost_connection_t *lc, sctp_assoc_t assoc_id);
 void rrc_gNB_process_f1_du_configuration_update(struct f1ap_gnb_du_configuration_update_s *conf_up, sctp_assoc_t assoc_id);
diff --git a/sib8.conf b/sib8.conf
new file mode 100644
index 0000000000..c9b5b4c6a8
--- /dev/null
+++ b/sib8.conf
@@ -0,0 +1,5 @@
+messageIdentifier=1112;
+serialNumber=FF00;
+dataCodingScheme=48;
+text=Hello this is a test warning message.;
+mode=0;
diff --git a/targets/PROJECTS/GENERIC-NR-5GC/CONF/gnb.sa.band78.fr1.106PRB.usrpb210.conf b/targets/PROJECTS/GENERIC-NR-5GC/CONF/gnb.sa.band78.fr1.106PRB.usrpb210.conf
index 7f750d7259..2bbc53ea22 100644
--- a/targets/PROJECTS/GENERIC-NR-5GC/CONF/gnb.sa.band78.fr1.106PRB.usrpb210.conf
+++ b/targets/PROJECTS/GENERIC-NR-5GC/CONF/gnb.sa.band78.fr1.106PRB.usrpb210.conf
@@ -17,10 +17,10 @@ gNBs =
 
     ////////// Physical parameters:
 
+
     do_CSIRS                                                  = 1;
     do_SRS                                                    = 1;
 
-    #uess_agg_levels = [0,1,2,2,1]
     servingCellConfigCommon = (
     {
  #spCellConfigCommon
@@ -30,7 +30,7 @@ gNBs =
 #  downlinkConfigCommon
     #frequencyInfoDL
       # this is 3600 MHz + 43 PRBs@30kHz SCS (same as initial BWP)
-      absoluteFrequencySSB                                             = 641280;
+      absoluteFrequencySSB                                             = 641280; 
       dl_frequencyBand                                                 = 78;
       # this is 3600 MHz
       dl_absoluteFrequencyPointA                                       = 640008;
@@ -49,7 +49,7 @@ gNBs =
         initialDLBWPsubcarrierSpacing                                   = 1;
       #pdcch-ConfigCommon
         initialDLBWPcontrolResourceSetZero                              = 12;
-        initialDLBWPsearchSpaceZero                                     = 0;
+        initialDLBWPsearchSpaceZero                                     = 2;
 
   #uplinkConfigCommon
      #frequencyInfoUL
@@ -217,7 +217,7 @@ security = {
   # preferred ciphering algorithms
   # the first one of the list that an UE supports in chosen
   # valid values: nea0, nea1, nea2, nea3
-  ciphering_algorithms = ( "nea0" );
+  ciphering_algorithms = ( "nea2" );
 
   # preferred integrity algorithms
   # the first one of the list that an UE supports in chosen
